src/backends/r1cs/mod.rs

```diff
@@ -0,0 +1,489 @@
+        label: Option<&'static str>,
+        value: Self::Field,
+        span: crate::constants::Span,
+    ) -> CellVar {
+        let x = self.new_internal_var(Value::Constant(value), span);
+        self.assert_eq_const(&x, value, span);
+
+        x
+    }
+
+    /// Final checks for generating the circuit.
+    /// todo: we might need to rethink about this interface
+    /// - private_input_indices are not needed in this r1cs backend.
+    /// - main_span could be better initialized with the backend, so it doesn't have to pass in here?
+    /// - we might just need the returned_cells argument, as the backend can record the public outputs itself?
+    fn finalize_circuit(
+        &mut self,
+        public_output: Option<crate::var::Var<Self::Field>>,
+        returned_cells: Option<Vec<CellVar>>,
+        _private_input_indices: Vec<usize>,
+        _main_span: crate::constants::Span,
+    ) -> crate::error::Result<()> {
+        // store the return value in the public input that was created for that
+        if let Some(public_output) = public_output {
+            let cvars = &public_output.cvars;
+
+            for (pub_var, ret_var) in cvars.clone().iter().zip(returned_cells.unwrap()) {
+                // replace the computation of the public output vars with the actual variables being returned here
+                let var_idx = pub_var.idx().unwrap();
+                let prev = &self.witness_vars[var_idx];
+                assert!(matches!(prev, Value::PublicOutput(None)));
+                self.witness_vars[var_idx] = Value::PublicOutput(Some(ret_var));
+            }
+        }
+
+        // for sanity check, we make sure that every cellvar created has ended up in an r1cs constraint
+        let mut written_vars = HashSet::new();
+        for constraint in &self.constraints {
+            for lc in constraint.as_array() {
+                for var in lc.terms.keys() {
+                    written_vars.insert(var.index);
+                }
+            }
+        }
+
+        // check if every cell vars end up being a cell var in the circuit or public output
+        for index in 0..self.witness_vars.len() {
+            if !written_vars.contains(&index) {
+                // check if outputs contains the cell var that has the same index
+                if !self.public_outputs.iter().any(|&x| x.index == index) {
+                    panic!("there's a bug in the circuit_writer, some cellvar does not end up being a cellvar in the circuit!");
+                }
+            }
+        }
+
+        self.finalized = true;
+
+        Ok(())
+    }
+
+    /// Generate the witnesses
+    /// This process should check if the constraints are satisfied.
+    fn generate_witness(
+        &self,
+        witness_env: &mut crate::witness::WitnessEnv<Self::Field>,
+    ) -> crate::error::Result<Self::GeneratedWitness> {
+        if !self.finalized {
+            panic!("the circuit is not finalized yet!");
+        }
```

Comment:
btw you can just write `assert!(self.finalized, "the circuit is not finalized yet!");`